[browser] [wasm] Make response streaming opt-out

campersau opened the pull request:
This PR makes browser HTTP client to support [streaming HTTP response](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Using_readable_streams#consuming_a_fetch_as_a_stream) by default. Until now, this feature was opt-in.

Because all evergreen browsers now support it.

This is compelling because it will make the `GetFromJsonAsAsyncEnumerable` streaming and lower memory consumption for large requests.

### breaking change 

This is a breaking change because the `response.Content.ReadAsStreamAsync()` is no longer `MemoryStream` but `BrowserHttpReadStream` which doesn't support synchronous operations like `Stream.Read(Span<Byte>)`. If your code uses synchronous operations, you can disable the feature or copy the stream into `MemoryStream` yourself.

### how to opt-out

If you need to disable it globally, you can use
- `<WasmEnableStreamingResponse>false</WasmEnableStreamingResponse>` in your project file
- set env variable `DOTNET_WASM_ENABLE_STREAMING_RESPONSE` 

To disable it for individual request, you can use existing 
- `request.Options.Set(new HttpRequestOptionsKey<bool>("WebAssemblyEnableStreamingResponse"), false);` 
- or `request.SetBrowserResponseStreamingEnabled(false)` extension in Blazor project

Related to #77904
Fixes to https://github.com/dotnet/runtime/issues/112442

campersau committed:
[browser] [wasm] Make response streaming opt-out

lewing reviewed (commented):
I support this in principle.  I'm concerned about how the breaking change will surface to people upgrading.  My assumption is that it at the moment it will only show up at runtime?

campersau commented:
Yes, it will only show up at runtime. Similar to Kestrels [`AllowSynchronousIO` defaulting to `false`](https://learn.microsoft.com/aspnet/core/fundamentals/servers/kestrel/options#synchronous-io).
The main issue I will have with dotnet/runtime types is that `ZipArchive` does not support async https://github.com/dotnet/runtime/issues/1541 and would throw.
An other workaround is to buffer the stream manually.

Could we add an app context switch to toggle this globally?

campersau commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -314,13 +314,7 @@ private HttpResponseMessage ConvertResponse()
-                }
-
-                responseMessage.Content = streamingResponseEnabled
+                responseMessage.Content = (_request.Options.TryGetValue(EnableStreamingResponse, out var streamingResponseEnabled) ? streamingResponseEnabled : true) && BrowserHttpInterop.SupportsStreamingResponse()
```
Since all browsers support response streaming, should we remove `BrowserHttpInterop.SupportsStreamingResponse`? 
- https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#browser_compatibility
- https://developer.mozilla.org/en-US/docs/Web/API/Response#browser_compatibility

pavelsavara commented:
> Could we add an app context switch to toggle this globally?

We could make it ILLink feature. 
Similar to `<InvariantGlobalization>true</InvariantGlobalization>` for example.

I'm not sure if making the breaking change is worth it. Most people don't stream large data.

I would stick to existing behavior as default.

MihaZupan commented:
FWIW this shouldn't be specific to just streaming large data.
A simple use case this affects can be streaming an AI chat response via `GetFromJsonAsAsyncEnumerable` or `SseParser`, where the preferred experience is that a partial response is displayed as soon as available.

lewing commented:
The GetFromJsonAsAsyncEnumerable case is very compelling to me given that the overhead of processing large json blobs is particularly painful in both memory and responsiveness in browser with a single thread.

campersau commented:
To me it looks like everyone supports this change. I guess it is still early to try it out in the previews then?

pavelsavara commented:
@campersau I think we are missing the global opt-out linker feature. Are you willing to improve the PR ?

pavelsavara commented:
I will work on adding global-opt out soon. @campersau I will take over this branch, ok ?

> ZipArchive does not support async

`BrowserHttpContent` doesn't support sync API already. 
So think this would not break anything new. Am I missing something ?

> Yes, it will only show up at runtime. 

If the developer didn't opt out, I can see few options:

A) at the start of request we can detect old browser and **fail the request** with some message hinting about possibilities how to opt out.
B) at the start of the runtime we can detect old browser and **fail the runtime startup**
C) we can detect old browser and resolve promise only once we received full response (no streaming) + issue warning to console

Thoughts ?

campersau commented:
> I will take over this branch, ok ?

Okay, I have tried to figure out how to do the linker opt out, but it is hard to find all the places which need to be updated.


> BrowserHttpContent doesn't support sync API already.
> So think this would not break anything new. Am I missing something ?

The `BrowserHttpContent` itself is only async but the `Stream` returned by `BrowserHttpContent.ReadAsStreamAsync` supports sync because it is just a `MemoryStream`:

https://github.com/dotnet/runtime/blob/bfc881417c984ee23c8903b2bd7aa42e75cbc800/src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs#L474

>If the developer didn't opt out, I can see few options:
>A) at the start of request we can detect old browser and fail the request with some message hinting about possibilities how to opt out.
>B) at the start of the runtime we can detect old browser and fail the runtime startup
>C) we can detect old browser and resolve promise only once we received full response (no streaming) + issue warning to console

I think all supported browsers support response streaming so I am not sure how you will detect it.
See https://github.com/dotnet/runtime/pull/111680#discussion_r1924823451

The `true` value here should be replaced by the global opt out value:
https://github.com/dotnet/runtime/pull/111680/files#diff-66ce677ded58c3a6a0639d321f4460bb0b2c4bcaeb3216ff3db4381937365bd8R317
Otherwise the developer can still opt out on a per request basis by setting `WebAssemblyEnableStreamingResponse` to `false`.

pavelsavara commented:
> Okay, I have tried to figure out how to do the linker opt out, but it is hard to find all the places which need to be updated.

Yes, no worries.

> I think all supported browsers support response streaming

Use the same detection we have there already. To detect **unsupported** browser and proceed or fail in meaningful way.

pavelsavara commented:
FYI @guardrex

guardrex commented:
@pavelsavara ... Is the new opt-out approach landing at Preview 2?

pavelsavara committed:
Merge branch 'main' into responsestreamingoptin

pavelsavara committed:
- WebAssemblyEnableStreamingResponse msbuild property

- System.Net.Http.WasmEnableStreamingResponse linker feature
- DOTNET_WASM_ENABLE_STREAMING_RESPONSE env variable
- tests & docs

pavelsavara committed:
Merge branch 'main' into responsestreamingoptin

pavelsavara committed:
- WebAssemblyEnableStreamingResponse msbuild property

- System.Net.Http.WasmEnableStreamingResponse linker feature
- DOTNET_WASM_ENABLE_STREAMING_RESPONSE env variable
- tests & docs

guardrex commented:
Thanks @pavelsavara. I'll react to the updated description on the docs PR next week, and I'll ping for docs PR review later.

pavelsavara committed:
Merge branch 'main' into responsestreamingoptin

pavelsavara committed:
fix linker ?

pavelsavara commented:
This passed both [FF](https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-111680-merge-93ad1c898e5048d881/WasmTestOnFirefox-ST-System.Net.Http.Functional.Tests/1/console.71d5a608.log?helixlogtype=result) and [chrome](https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-111680-merge-56747c2b7e2b4cfcba/WasmTestOnChrome-ST-System.Net.Http.Functional.Tests/1/console.5920164f.log?helixlogtype=result) tests.

campersau commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpInterop.cs:
```diff
@@ -11,11 +11,26 @@ namespace System.Net.Http
     internal static partial class BrowserHttpInterop
     {
+        private static bool? _SupportsStreamingRequest;
+        private static bool? _SupportsStreamingResponse;
```
Since this is now cached in .NET we don't need to cache it in JS anymore. Does not hurt though and might still be relevant for MT.

pavelsavara replied:
I removed it there, thanks. I believe that the webworker has the same feature set as UI thread. 
If not, having static cache in C# side would be wrong.

pavelsavara commented on src/libraries/System.Net.Http/src/ILLink/ILLink.Substitutions.xml:
```diff
@@ -4,4 +4,9 @@
   </assembly>
+  <assembly fullname="System.Net.Http" feature="System.Net.Http.WebAssemblyEnableStreamingResponse" featurevalue="false">
+    <type fullname="System.Net.Http.BrowserHttpController">
+      <method signature="System.Boolean get_FeatureEnableStreamingResponse()" body="stub" value="false" />
```
This means: if you set `System.Net.Http.WebAssemblyEnableStreamingResponse` to `false` then ILLink will stub the `get_FeatureEnableStreamingResponse` to return `false` instead of reading the config/env.

We don't want to also hardcode `true` do we ?

javiercn replied:
Is the behavior here the same in development as in publish?

We have something similar in Blazor routing and we had to do extra work for the development case
 https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/Properties/ILLink.Substitutions.xml#L6-L8 for routing constraints, and it doesn't work in development unless you specifically set the AppContextSwitch manually

https://github.com/dotnet/aspnetcore/blob/f25dc7be397b496cc1c71f8720b2b7d67fb18649/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs#L98-L112

pavelsavara committed:
Update src/mono/wasm/features.md

Co-authored-by: campersau <buchholz.bastian@googlemail.com>

pavelsavara committed:
feedback

javiercn commented on src/mono/wasm/features.md:
```diff
@@ -55,6 +55,14 @@ Because web browsers do not expose direct access to sockets, we are unable to pr
 
 A prominent limitation is that your application must obey `Cross-Origin Resource Sharing` (CORS) rules in order to perform network requests successfully - see [CORS on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) for more information.
 
+Since Net 10 Preview 3 the HTTP client supports [streaming HTTP response](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Using_readable_streams#consuming_a_fetch_as_a_stream) by default because all evergreen browsers now support it.
+
+This is a breaking change because the `response.Content.ReadAsStreamAsync()` is no longer `MemoryStream` but `BrowserHttpReadStream` which doesn't support synchronous operations like `Stream.Read(Span<Byte>)`. If your code uses synchronous operations, you can disable the feature or copy the stream into `MemoryStream` yourself.
+
+If you need to disable it, you can use `<WasmEnableStreamingResponse>false</WasmEnableStreamingResponse>` or `DOTNET_WASM_ENABLE_STREAMING_RESPONSE` env variable to do it for all HTTP requests.
+
+Or you can use `request.Options.Set(new HttpRequestOptionsKey<bool>("WebAssemblyEnableStreamingResponse"), false);` for individual request.
+
```
Does it make sense to unconditionally enable this on well-known APIs?

If you are forced to turn it off for some reason (using a library that makes HttpRequests that read synchronously and which you don't have control of) then you have to turn it off globally, and that means you stop being able to use things like the async enumerable APIs. 

Does it make sense for those APIs to turn the feature on unconditionally in their implementation?

pavelsavara replied:
I'm thinking about it this way. If the new behavior gives you trouble, you can use the `<WasmEnableStreamingResponse>false</WasmEnableStreamingResponse>` to get to previous behavior.

Also, now you can choose if enabled by default or disabled by default gives you less trouble.

> Does it make sense for those APIs to turn the feature on unconditionally in their implementation?

Those who know enough and have control over it already did in Net8. This PR is "let's make the better default for everyone".

javiercn replied:
Great, thanks for the clarification.

pavelsavara commented:
This is now blocked by https://github.com/dotnet/runtime/issues/97449

pavelsavara commented:
@lewing I think we should merge it into P3 anyway. Quite likely dev-loop users do have latest browsers. 
So they would be only hit by the MemoryStream sync APIs problem and we would hear feedback sooner.

pavelsavara commented:
/ba-g CI fails are unrelated

pavelsavara commented:
Hmm, my bad, I overlooked HttpStreamingDisabledBy_WasmEnableStreamingResponse_InProject [fail](https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-111680-merge-8d053d63c9bd4dd082/WasmTestOnChrome-ST-System.Runtime.InteropServices.JavaScript.Tests/1/console.8cf854a9.log?helixlogtype=result)

https://github.com/dotnet/runtime/issues/113628
